Java API Generics Phase 3.

andrewcmyers opened the pull request:
- Added the utility Tensors class.
- Updated tests to use it.
- Updated scripts for generating Tensors.java and the types directory.
  Note that these are still run manually, but remain helpful because
  maintaining so many methods and their documentation is a headache.
- Added missing checking for attempts to create tensors from arrays of boxed
  primitives, and a test case.

andrewcmyers committed:
Java API Generics Phase 3.

- Added the utility Tensors class.
- Updated tests to use it.
- Updated scripts for generating Tensors.java and the types directory.
  Note that these are still run manually, but remain helpful because
  maintaining so many methods and their documentation is a headache.
- Added missing checking for attempts to create tensors from arrays of boxed
  primitives, and a test case.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

andrewcmyers commented:
@asimshankar Java Generics Phase 3 PR started.

Some issues to consider:
- Do we want to change the type parameter for strings?  It's genuinely unclear to me how often encoding mistakes will happen and how much TFString would actually help. The idea of getting more opinions sounds good; I can go either way.
- Should the scripts for code generation be ported to some other language? They are not part of the build process, so maybe it's not so critical.
- We will probably want parallel methods for creating constant Tensors, in `op.Constant`, folding together `Constant.createWithTensor` + `Tensors.create`. I'm thinking that can go into a later PR.

sb2nov commented:
Jenkins, test this please.

asimshankar commented:
Thanks for the PR. A few quick things:

- RE: scripts to generate: Yes, we should port to another language (or heck, use an `AnnotationProcessor` :)). But can defer that for now because, as you point out, the generation scripts are not part of the build process (and have to be manually invoked). However, I didn't see the generated `Tensors.java` in this PR, shouldn't that be included too?

- RE: Merging `Constant.createWithTensor` and `Tensors.create` sounds like a great idea, but again, something that we can do later on.

Please update the PR with the generated `Tensors.java`.
Thanks

andrewcmyers committed:
Oops. This is the generated file!

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Tensor.java:
```diff
@@ -637,11 +638,25 @@ private static void throwExceptionIfNotByteOfByteArrays(Object array) {
    * uint8} and {@code string}, with the latter being the default interpretation.
    */
-  private static DataType dataTypeOf(Object o) {
+  private static Class<?> baseObjType(Object o) {
```
Javadoc above needs to be updated.
(Or you can just remove it - 5 line private helper methods don't really need a Javadoc :)

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Tensors.java:
```diff
@@ -0,0 +1,427 @@
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/** Utility class offering convenience methods for creating Tensors in a type-safe way. */
```
Optional: How about wording this as:

```java
/** Type-safe factory methods for creating {@link Tensor} objects. */
```

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Tensors.java:
```diff
@@ -0,0 +1,427 @@
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/** Utility class offering convenience methods for creating Tensors in a type-safe way. */
+public class Tensors {
```
Nit: `public final class`?

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/types/UInt8.java:
```diff
@@ -17,5 +17,8 @@
 /** Represents an 8-bit unsigned integer. */
 public class UInt8 {
-  private UInt8() {}
+  private byte value;
```
I didn't quite follow the point of these private constructors and value, could you elaborate?

andrewcmyers replied:
This was setting up to support a future generic `scalarValue()` accessor. But I was getting ahead of myself...

asimshankar reviewed (commented):
A few minor suggestions, otherwise looks great. Thanks!

FYI @kbsriram @karllessard

andrewcmyers committed:
Addressed Asim's comments.
- made Tensors final
- got rid of unused state in UInt8 objects
- tuned up various javadoc comments

asimshankar commented:
Thanks @andrewcmyers ! Much appreciated.

asimshankar commented:
Jenkins, test this please

asimshankar commented:
Jenkins, test this please
